parquet-column/src/main/java/org/apache/parquet/column/values/bitpacking/ParquetReadRouter.java

```diff
@@ -0,0 +1,133 @@
+    BytePacker packerVector = Packer.LITTLE_ENDIAN.newBytePackerVector(bitWidth);
+    int valueIndex = 0;
+    int byteIndex = 0;
+    int unpackCount = packerVector.getUnpackCount();
+    int inputByteCountPerVector = packerVector.getUnpackCount() / BITS_PER_BYTE * bitWidth;
+    int totalByteCount = currentCount * bitWidth / BITS_PER_BYTE;
+    int totalByteCountVector = totalByteCount - BYTES_PER_VECTOR_512;
+    ByteBuffer buffer = in.slice(totalByteCount);
+    if (buffer.hasArray()) {
+      for (; byteIndex < totalByteCountVector; byteIndex += inputByteCountPerVector, valueIndex += unpackCount) {
+        packerVector.unpackValuesUsingVector(buffer.array(), buffer.arrayOffset() + buffer.position() + byteIndex, currentBuffer, valueIndex);
+      }
+      // If the remaining bytes size <= {BYTES_PER_512VECTOR}, the remaining bytes are unpacked by packer
+      for (; byteIndex < totalByteCount; byteIndex += bitWidth, valueIndex += NUM_VALUES_TO_PACK) {
+        packer.unpack8Values(buffer.array(), buffer.arrayOffset() + buffer.position() + byteIndex, currentBuffer, valueIndex);
+      }
+    } else {
+      for (; byteIndex < totalByteCountVector; byteIndex += inputByteCountPerVector, valueIndex += unpackCount) {
+        packerVector.unpackValuesUsingVector(buffer, buffer.position() + byteIndex, currentBuffer, valueIndex);
+      }
+      for (; byteIndex < totalByteCount; byteIndex += bitWidth, valueIndex += NUM_VALUES_TO_PACK) {
+        packer.unpack8Values(buffer, buffer.position() + byteIndex, currentBuffer, valueIndex);
+      }
+    }
+  }
+
+  // Call the method directly if your computer system doesn't contain avx512vbmi and avx512_vbmi2 CPU Flags
+  public static void readBatch(int bitWidth, ByteBufferInputStream in, int currentCount, int[] currentBuffer) throws EOFException {
+    BytePacker packer = Packer.LITTLE_ENDIAN.newBytePacker(bitWidth);
+    int valueIndex = 0;
+    while (valueIndex < currentCount) {
+      ByteBuffer buffer = in.slice(bitWidth);
+      packer.unpack8Values(buffer, buffer.position(), currentBuffer, valueIndex);
+      valueIndex += NUM_VALUES_TO_PACK;
+    }
+  }
+
+  private static VectorSupport getSupportVectorFromCPUFlags() {
+    try {
+      String os = System.getProperty("os.name");
+      if (os == null || !os.toLowerCase().startsWith("linux")) {
+        return VectorSupport.NONE;
+      }
+      List<String> allLines = Files.readAllLines(Paths.get("/proc/cpuinfo"), StandardCharsets.UTF_8);
+      for (String line : allLines) {
+        if (line != null && line.startsWith("flags")) {
+          int index = line.indexOf(":");
+          if (index < 0) {
+            continue;
+          }
+          line = line.substring(index + 1);
+          Set<String> flagsSet = Arrays.stream(line.split(" ")).collect(Collectors.toSet());
+          if (flagsSet.contains("avx512vbmi") && flagsSet.contains("avx512_vbmi2")) {
+            return VectorSupport.VECTOR_512;
+          }
+        }
+      }
```

Comment:
@gszadovszky avx512vbmi and avx512_vbmi2 are CPU Flags which contains some instruction set. The code of this PR  has used the instruction set. Without them, we cannot get benefit from Java Vector API.

Reply:
@jiangjiguang, I understand how it works. What I would have expected from the Vector API is to implement some help to check on the CPU flags.

Reply:
+1 for the request for official support to reduce maintenance burden.

Reply:
@jatin-bhateja Can you help give an answer ? thanks

Reply:
> @jiangjiguang, I understand how it works. What I would have expected from the Vector API is to implement some help to check on the CPU flags.

Hi @gszadovszky , vector APIs are just in time compiled into efficient instruction sequence, this instruction sequence can vary based on CPU features set of the target platform,  by introducing additional feature checks for vectorized unpacking we are trying to prevent any perf degradations over non-X86 platforms, over the time these features checks may be relaxed once we quantify the performance improvements on other targets.

One can use JVM option -XX:+PrintIntrinsics to determine if  VectorAPI used in the source is intrinsified for a particular target.

Reply:
@jatin-bhateja, thanks for the explanation. This is fine to check CPU flags. I was wondering, though if there is an existing 3rd party implementation to check things like that so we don't take the burden of maintaining this code.

Reply:
>  I was wondering, though if there is an existing 3rd party implementation to check things like that

Hi @gszadovszky , Any implementation using VectorAPI will need some fine tuning since algorithm may not be performant if any API does not get intrinsified for a specified target.

>  burden of maintaining this code.

I am not aware of any 3rd party java based library which vectorizes parquet-mr bit packing.  Intel has also contributed a VectorAPI base version of [PForR Java library](https://github.com/lemire/JavaFastPFOR/pull/51) which does integral bit packing which is other encoding scheme being used in parquet-mr but its a different format.

![PFOR_VS_PARQUET_FORMAT](https://user-images.githubusercontent.com/59989778/220858640-26c07f03-9415-4152-834e-1bc90f02c554.png)

Reply:
Sorry, @jatin-bhateja, I was misleading. I was talking about this very code about checking the CPU flags not the whole vectorization. I accept we need to check the CPU flags. I only think that the way we are doing is quite low level. It might be implemented already by a 3rd party that might support other OS etc.

Reply:
@jiangjiguang , can you please respond to @gszadovszky  query.

Reply:
@gszadovszky Even though the java vector API can check the CPU Flags(low level check), but if users use java vector api on computer which dose not contain avx512vbmi && avx512_vbmi2, the java vector API how to do ?  throw exception?  fall back ?  

Besides,  the CPU Flags are becoming more and more,  and  Java vector api is also increasing. maybe Java vector api dose not know which CPU Flags  the users need. 
I think  users should have already known their computer(including CPU Flags). They can decide whether to  enable parquet java17 vector optimization. 

For example, velox(https://github.com/facebookincubator/velox) shows only what CPU Flags needed.
<img width="1085" alt="image" src="https://user-images.githubusercontent.com/12368495/220878686-53b6eec5-89e9-4958-8802-ff5fd97076bd.png">

Reply:
@gszadovszky I got it. As far as I know, there is no such library available for java. I think we can try to achieve implement a library like that with JNI in the future. But there is no time line. If I have finished it, I will update here.